annotated_test.go

```diff
@@ -1450,3 +1455,59 @@ func TestHookAnnotationFailures(t *testing.T) {
 		})
 	}
 }
+
+func TestHookAnnotationFunctionFlexibility(t *testing.T) {
+	type A interface{}
+
+	called := atomic.NewBool(false)
+	table := []struct {
+		name       string
+		annotation interface{}
+	}{
+		{
+			name: "without error return",
+			annotation: fx.Annotate(
+				func() A {
+					return nil
+				},
+				fx.OnStart(func(context.Context) {
+					called.Store(true)
+				}),
+			),
+		},
+		{
+			name: "without context param",
+			annotation: fx.Annotate(
+				func() A { return nil },
+				fx.OnStart(func() error {
+					called.Store(true)
+					return nil
+				}),
+			),
+		},
+		{
+			name: "without context param or error return",
+			annotation: fx.Annotate(
+				func() A { return nil },
+				fx.OnStart(func() {
+					called.Store(true)
+				}),
+			),
+		},
+	}
+
+	for _, tt := range table {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			called.Store(false)
+
+			opts := fx.Options(
+				fx.Provide(tt.annotation),
+				fx.Invoke(func(A) {}),
+			)
```

Comment:
optional:
we can avoid sharing `called` between the tests by putting `called` into the container,
and making that the parameter the OnStart functions are accepting.

```
var called atomic.Bool
opts := fx.Options(
  fx.Supply(&called)
  // ...
)
```

And in the test,

```
fx.OnStart(func(called *atomic.Bool) {
  called.Store(true)
})
```

Reply:
Ah good call, thanks!